File: scripts/ghidra/export_all_functions.py
# Ghidra headless post-script: decompiles every function in the current
# program above a minimum size (to skip obvious libc/thunk noise) and
# writes each to its own .c file, named by entry-point address (no
# symbol names survive stripping, so address is the only stable label).
#
# Run via analyzeHeadless with:
#   -postScript export_all_functions.py <output_dir> <min_size_bytes>
#
# Jython-compatible (Ghidra's classic scripting API) - avoid Python3-only
# syntax.
#
# @category IRSS

import os
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

args = getScriptArgs()
if len(args) < 2:
    print("Usage: export_all_functions.py <output_dir> <min_size_bytes>")
    exit(1)

out_dir = args[0]
min_size = int(args[1])

if not os.path.exists(out_dir):
    os.makedirs(out_dir)

monitor = ConsoleTaskMonitor()
decomp = DecompInterface()
decomp.openProgram(currentProgram)

fm = currentProgram.getFunctionManager()
functions = fm.getFunctions(True)  # True = forward order

total = 0
exported = 0
skipped_thunk = 0
skipped_external = 0
skipped_small = 0
skipped_failed = 0

for func in functions:
    total += 1

    if func.isThunk():
        skipped_thunk += 1
        continue
    if func.isExternal():
        skipped_external += 1
        continue

    body_size = func.getBody().getNumAddresses()
    if body_size < min_size:
        skipped_small += 1
        continue

    result = decomp.decompileFunction(func, 60, monitor)
    if not result.decompileCompleted():
        skipped_failed += 1
        print("FAILED to decompile %s: %s" % (func.getEntryPoint(), result.getErrorMessage()))
        continue

    code = result.getDecompiledFunction().getC()
    addr_str = str(func.getEntryPoint())
    out_path = os.path.join(out_dir, "FUN_%s.c" % addr_str)
    f = open(out_path, "w")
    f.write(code)
    f.close()
    exported += 1

print("=== export_all_functions summary ===")
print("Total functions seen:   %d" % total)
print("Exported:               %d" % exported)
print("Skipped (thunk):        %d" % skipped_thunk)
print("Skipped (external):     %d" % skipped_external)
print("Skipped (< min_size):   %d" % skipped_small)
print("Skipped (decomp fail):  %d" % skipped_failed)
print("Output dir: %s" % out_dir)

File: scripts/ghidra/ExportAllFunctions.java
// Ghidra headless post-script: decompiles every function in the current
// program above a minimum size (to skip obvious libc/thunk noise) and
// writes each to its own .c file, named by entry-point address (no
// symbol names survive stripping, so address is the only stable label).
//
// Run via analyzeHeadless with:
//   -postScript ExportAllFunctions.java <base_output_dir> <min_size_bytes>
//
// Java, not the old Jython .py scripts (Ghidra 12.x dropped Jython
// support in favor of PyGhidra, which needs a separate Python runtime
// we haven't set up) - a compiled GhidraScript works with zero extra
// setup in any Ghidra version.
//
// Output is written to <base_output_dir>/<program's project path>/, so
// -process with -recursive can safely sweep up multiple same-named
// programs (e.g. two files both literally called "httpd" in different
// project folders) in one invocation without their output colliding -
// local Ghidra projects have no way to scope -process to one folder,
// only by filename.
//
// @category IRSS

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;

import java.io.File;
import java.io.FileWriter;

public class ExportAllFunctions extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 2) {
            println("Usage: ExportAllFunctions.java <output_dir> <min_size_bytes>");
            return;
        }
        String baseOutDirPath = args[0];
        int minSize = Integer.parseInt(args[1]);

        // e.g. "/netgear/httpd" -> "netgear_httpd", "/netgear/patched/httpd" -> "netgear_patched_httpd"
        String programPath = currentProgram.getDomainFile().getPathname();
        String subDirName = programPath.replaceAll("^/+", "").replaceAll("[/\\\\]+", "_");
        File dir = new File(baseOutDirPath, subDirName);
        if (!dir.exists()) {
            dir.mkdirs();
        }
        String outDirPath = dir.getAbsolutePath();

        DecompInterface decomp = new DecompInterface();
        DecompileOptions options = new DecompileOptions();
        options.grabFromProgram(currentProgram);
        decomp.setOptions(options);
        // Match the GUI Decompile panel's default simplification style -
        // a bare DecompInterface without this explicitly set leaves
        // pointer references as raw DAT_<addr> instead of resolving them
        // to inline string literals, even when the underlying data is
        // already typed/labeled as a string by prior analysis.
        decomp.setSimplificationStyle("decompile");
        decomp.openProgram(currentProgram);

        FunctionIterator functions = currentProgram.getFunctionManager().getFunctions(true);

        int total = 0;
        int exported = 0;
        int skippedThunk = 0;
        int skippedExternal = 0;
        int skippedSmall = 0;
        int skippedFailed = 0;

        for (Function func : functions) {
            if (monitor.isCancelled()) {
                break;
            }
            total++;

            if (func.isThunk()) {
                skippedThunk++;
                continue;
            }
            if (func.isExternal()) {
                skippedExternal++;
                continue;
            }

            long bodySize = func.getBody().getNumAddresses();
            if (bodySize < minSize) {
                skippedSmall++;
                continue;
            }

            DecompileResults results = decomp.decompileFunction(func, 60, monitor);
            if (!results.decompileCompleted()) {
                skippedFailed++;
                println("FAILED to decompile " + func.getEntryPoint() + ": " + results.getErrorMessage());
                continue;
            }

            String code = results.getDecompiledFunction().getC();
            String addrStr = func.getEntryPoint().toString();
            File outFile = new File(dir, "FUN_" + addrStr + ".c");
            FileWriter fw = new FileWriter(outFile);
            fw.write(code);
            fw.close();
            exported++;
        }

        println("=== ExportAllFunctions summary ===");
        println("Total functions seen:   " + total);
        println("Exported:               " + exported);
        println("Skipped (thunk):        " + skippedThunk);
        println("Skipped (external):     " + skippedExternal);
        println("Skipped (< min_size):   " + skippedSmall);
        println("Skipped (decomp fail):  " + skippedFailed);
        println("Output dir: " + outDirPath);
    }
}
